plan/eliminate_projection.go

```diff
@@ -14,50 +14,215 @@
-		child.SetSchema(plan.Schema())
-		RemovePlan(p)
-		p = EliminateProjection(child)
-	}
-	children := make([]Plan, 0, len(p.Children()))
-	for _, child := range p.Children() {
-		children = append(children, EliminateProjection(child.(PhysicalPlan)))
+// canLogicalProjectionBeEliminated checks if a PROJECTION in a logical plan can be eliminated.
+// Projection p can be eliminated when all the following conditions are satisfied:
+// 1. it's not the root operator
+// 2. canProjectionBeEliminated(p) returns true
+func canLogicalProjectionBeEliminated(p *Projection) bool {
+	if len(p.Parents()) == 0 {
+		return false
 	}
-	p.SetChildren(children...)
-	return p
+	return canProjectionBeEliminated(p)
 }
 
-// projectionCanBeEliminated checks if a PROJECTION operator can be eliminated.
-// PROJECTION operator can be eliminated when meet the following conditions at the same time:
+// canProjectionBeEliminated checks if a PROJECTION in a physical plan can be eliminated.
+// Projection p can be eliminated when all the following conditions are satisfied:
 // 1. fields of PROJECTION are all columns
 // 2. fields of PROJECTION are just the same as the schema of the child operator (including order, amount, etc.).
 // expressions like following cases can not be eliminated:
 // "SELECT b, a FROM t",
 // or "SELECT c AS a, c AS b FROM t WHERE d = 1",
 // or "SELECT t1.a, t2.b, t1.b, t2.a FROM t1, t2 WHERE t1.a < 0 AND t2.b > 0".
-func projectionCanBeEliminated(p *Projection) bool {
-	child := p.children[0].(PhysicalPlan)
+func canProjectionBeEliminated(p *Projection) bool {
+	child := p.Children()[0]
 	if p.Schema().Len() != child.Schema().Len() {
 		return false
 	}
+	childCols := child.Schema().Columns
 	for i, expr := range p.Exprs {
-		col, ok := expr.(*expression.Column)
-		if !ok {
-			return false
-		}
-		if col.FromID != child.Schema().Columns[i].FromID || col.Position != child.Schema().Columns[i].Position {
+		projCol, ok := expr.(*expression.Column)
+		if !ok || !projCol.Equal(childCols[i], nil) {
 			return false
 		}
 	}
 	return true
 }
+
+func resolveColumnAndReplace(origin *expression.Column, replace map[string]*expression.Column) {
+	dst := replace[string((*origin).HashCode())]
+	if dst != nil {
+		*origin = *dst
+	}
+}
+
+func resolveExprAndReplace(origin expression.Expression, replace map[string]*expression.Column) {
+	switch expr := origin.(type) {
+	case *expression.Column:
+		resolveColumnAndReplace(expr, replace)
+	case *expression.ScalarFunction:
+		for _, arg := range expr.GetArgs() {
+			resolveExprAndReplace(arg, replace)
+		}
+	}
+}
+
+func eliminateRootProjection(p Plan) Plan {
+	plan, ok := p.(*Projection)
+	if ok && canProjectionBeEliminated(plan) {
+		child := plan.Children()[0]
+		childColumns := child.Schema().Columns
+		for i, parentColumn := range plan.Schema().Columns {
+			childColumns[i].ColName = parentColumn.ColName
+		}
+		RemovePlan(p)
+		return child
+	}
+	return p
+}
+
+type nonRootProjectionEliminater struct {
```

Comment:
After refactoring, projection elimination is done in both logical and physical optimization, but they are different.

In logical optimization, we only eliminate the no-root projections, leave the root projection, if exists and can be eliminated, to be eliminated in the physical optimization

I hope to use these abnormal function and struct names, which are `nonRootProjectionEliminater` and `eliminateRootProjection`, to inform people the difference between them.